napari/config.py

```diff
@@ -0,0 +1,689 @@
+    and/or lists.
+
+    Parameters
+    ----------
+    config : dict, iterable, or str
+        Input object to search for environment variables
+
+    Returns
+    -------
+    config : same type as input
+
+    Examples
+    --------
+    >>> expand_environment_variables({'x': [1, 2, '$USER']})
+    {'x': [1, 2, 'my-username']}
+    """
+    if isinstance(config, Mapping):
+        return {k: expand_environment_variables(v) for k, v in config.items()}
+    elif isinstance(config, str):
+        return os.path.expandvars(config)
+    elif isinstance(config, (list, tuple, builtins.set)):
+        return type(config)([expand_environment_variables(v) for v in config])
+    else:
+        return config
+
+
+deprecations = {
+    "fuse_ave_width": "optimization.fuse.ave-width",
+    "fuse_max_height": "optimization.fuse.max-height",
+    "fuse_max_width": "optimization.fuse.max-width",
+    "fuse_subgraphs": "optimization.fuse.subgraphs",
+    "fuse_rename_keys": "optimization.fuse.rename-keys",
+    "fuse_max_depth_new_edges": "optimization.fuse.max-depth-new-edges",
+}
+
+
+def check_deprecations(key: str, deprecations: dict = deprecations):
+    """Check if the provided value has been renamed or removed.
+
+    Parameters
+    ----------
+    key : str
+        The configuration key to check
+    deprecations : Dict[str, str]
+        The mapping of aliases
+
+    Examples
+    --------
+    >>> deprecations = {"old_key": "new_key", "invalid": None}
+    >>> check_deprecations("old_key", deprecations=deprecations)
+    UserWarning: Configuration key "old_key" has been deprecated. Please use "new_key" instead.
+
+    >>> check_deprecations("invalid", deprecations=deprecations)
+    Traceback (most recent call last):
+        ...
+    ValueError: Configuration value "invalid" has been removed
+
+    >>> check_deprecations("another_key", deprecations=deprecations)
+    'another_key'
+
+    Returns
+    -------
+    new: str
+        The proper key, whether the original (if no deprecation) or the aliased
+        value
+    """
+    if key in deprecations:
+        new = deprecations[key]
+        if new:
+            warnings.warn(
+                'Configuration key "{}" has been deprecated. '
+                'Please use "{}" instead'.format(key, new)
+            )
+            return new
+        else:
+            raise ValueError(
+                'Configuration value "{}" has been removed'.format(key)
+            )
+    else:
+        return key
+
+
+def sync(config=config, destination=None, lock=config_lock):
+    dirty = config.pop('_dirty', None)
+    if not dirty:
+        return
+
+    if destination is None:
+        destination = PATH
+
+    if os.path.isdir(destination):
+        destination = os.path.join(destination, 'last_session.yaml')
+
+    with lock:
+        # TODO: should we also read from the file and merge new keys?
+        with open(destination, 'w') as f:
+            yaml.safe_dump(config, f)
```

Comment:
I added this, (not in dask).  Dask seems to only _read in_ config, but never write it out.  this function writes the current config to `last_session.yaml` in the default PATH.  (happy to change the name).  _all_ yaml files in PATH get read on launch.  we can refine this later.